### PR TITLE
General optimizations

### PR DESCRIPTION
**Pull request by cmleinz**
Thanks for writing this library, very cool!

These are just a few optimizations to avoid allocating where it is not needed, and to avoid certain costly operations (i.e. reversing a large vector)

If you think it is worth it in the future to add another dependency, another optimization would be to use a small stack allocated string array from [arrayvec](https://docs.rs/arrayvec/latest/arrayvec/struct.ArrayString.html) to prevent heap allocating format strings for each log entry. This would mean there would need to be some limit on the size of the displayed log entry, so it's not without some trade offs.

### CODE REVIEW

**Comment by RegenJacob** [comment]
Great work! Thanks for your pull request 

I was aware that performance wasn't great so this will be quite an upgrade I assume.

I will test in a few hours

File: src/lib.rs [hunk 1]
```diff
@@ -144,29 +140,20 @@ impl LoggerUi {
...
-                            ui.colored_label(Color32::YELLOW, string_format);
-                        }
-                        log::Level::Error => {
-                            ui.colored_label(Color32::RED, string_format);
-                        }
-                        _ => {
-                            ui.label(string_format);
-                        }
-                    }
+                    let _ = match level {
```
Comment (RegenJacob) [inline src/lib.rs]: Is there a reason for the `let _`?

Comment (RegenJacob) [reply src/lib.rs]: I'm just curios if it actually matters or if it's just a formatting choice

**Comment by RegenJacob** [comment]
> If you think it is worth it in the future to add another dependency, another optimization would be to use a small stack allocated string array from [arrayvec](https://docs.rs/arrayvec/latest/arrayvec/struct.ArrayString.html) to prevent heap allocating format strings for each log entry. This would mean there would need to be some limit on the size of the displayed log entry, so it's not without some trade offs.

This could be added as an optional feature flag later on. I'm not sure about limiting the size of the log as it could be important in some environments (e.g. game engines).

File: src/lib.rs (hunk 1 above)
Comment (cmleinz) [reply src/lib.rs]: You're totally right, this is purely a formatting thing, feel free to omit.

Comment (RegenJacob) [reply src/lib.rs]: I see, in this case I'll omit this. 

thanks!

**Comment by cmleinz** [comment]
> > If you think it is worth it in the future to add another dependency, another optimization would be to use a small stack allocated string array from [arrayvec](https://docs.rs/arrayvec/latest/arrayvec/struct.ArrayString.html) to prevent heap allocating format strings for each log entry. This would mean there would need to be some limit on the size of the displayed log entry, so it's not without some trade offs.
> 
> This could be added as an optional feature flag later on. I'm not sure about limiting the size of the log as it could be important in some environments (e.g. game engines).

That makes sense to me. As an alternative, I think it's probably worth it to cache the results of the list of displayed strings so that each frame does not construct the full list of format strings, but instead can pass references to each display string. That could be a separate PR though

**Comment by RegenJacob** [comment]
> > > If you think it is worth it in the future to add another dependency, another optimization would be to use a small stack allocated string array from [arrayvec](https://docs.rs/arrayvec/latest/arrayvec/struct.ArrayString.html) to prevent heap allocating format strings for each log entry. This would mean there would need to be some limit on the size of the displayed log entry, so it's not without some trade offs.
> > 
> > 
> > This could be added as an optional feature flag later on. I'm not sure about limiting the size of the log as it could be important in some environments (e.g. game engines).
> 
> That makes sense to me. As an alternative, I think it's probably worth it to cache the results of the list of displayed strings so that each frame does not construct the full list of format strings, but instead can pass references to each display string. That could be a separate PR though

Yes, I think that this pull request is good to go.

**Pull request merged**